### PR TITLE
Compatibility with PHP 8

### PR DESCRIPTION
**Pull request by remicollet**
Seems quite huge... sorry

Test suite is mostly OK, 4 failed tests, mostly related to error => exception change

### CODE REVIEW

**Comment by remicollet** [comment]
Perhaps will have be simpler to create a "php8" directory...

**Comment by remicollet** [comment]
Notice, this also includes a fix for 7.4 where write_property have to return a zval (instead of void)

**Comment by remicollet** [comment]
Local Build ok  with 7.3.22, 7.4.10 and 8.0.0beta3, waiting for travis ;)

**Comment by remicollet** [comment]
Now test suite passes with 7.0, 7.3 and 8.0 :+1:

File: tests/041.solrobject_illegal_operation.phpt [hunk 1]
```diff
@@ -48,9 +48,7 @@ Exception 1006: SolrObject instances are read-only. Properties cannot be added,
 
 Warning: main(): Attempting to set value for [newprop2_dimension_access] property in a SolrObject instance in %s on line %d
 Exception 1006: SolrObject instances are read-only. Properties cannot be added, updated or removed.
-
-Warning: main(): Attempting to remove [responseHeader] property in a SolrObject instance in %s on line %d
```
Comment (remicollet) [inline tests/041.solrobject_illegal_operation.phpt:52]: Don't know why need, unrealted to PHP 8.
See some failure about this in travis, not always...

File: tests/019.solrclient_getdebug.phpt [hunk 2]
```diff
@@ -28,6 +28,7 @@ sort($lines);
 foreach ( $lines as $line) {
 	if ($line == '<?xml version="1.0" encoding="UTF-8"?>' ||
 		FALSE !== strpos($line, 'Trying') ||
+		0 === strpos($line, 'Mark bundle') ||
```
Comment (remicollet) [inline tests/019.solrclient_getdebug.phpt:31]: Additional line displayed, don't know why, please check

Comment (remicollet) [reply tests/019.solrclient_getdebug.phpt:31]: `Mark bundle as not supporting multiuse`

**Comment by remicollet** [comment]
Travis seems happy, so ready for review

**Comment by Jan-E** [comment]
config.w32 needs adjustments because of 
https://github.com/php/pecl-search_engine-solr/blob/master/config.w32#L80
```
ERROR: Cannot match any known PHP version with 'php8.dll'
The system cannot find the file specified.

Microsoft (R) Program Maintenance Utility Version 14.27.29111.0
Copyright (C) Microsoft Corporation.  All rights reserved.

NMAKE : fatal error U1073: don't know how to make 'php_solr.dll'
Stop.
```

**Comment by Jan-E** [comment]
This patch (against https://github.com/remicollet/pecl-search_engine-solr/tree/issue-php8) makes it compile with PHP 8.0.0beta3 and VS16:
```
diff --git a/config.w32 b/config.w32
index 1fa62fa..cd95096 100644
--- a/config.w32
+++ b/config.w32
@@ -59,10 +59,6 @@ if(PHP_SOLR != 'no')
 			PHP_SOLR = "no";
 		}
 	}
-	if(PHP_JSON == 'no' || !ADD_EXTENSION_DEP('solr', 'json')) {
-		WARNING('solr was not enabled; json is not enabled');
-		PHP_SOLR = "no";
-	}
 	if(PHP_SOLR != 'no') {
 		if(PHP_SOLR_DEBUG == 'yes') {
 			AC_DEFINE('SOLR_DEBUG', 1, 'Solr debugging set to on');
@@ -77,7 +73,7 @@ if(PHP_SOLR != 'no')
             then set it back*/
         if (null != dll.match(/^php5/)) {
             configure_module_dirname = configure_module_dirname + "\\src\\php5";
-        } else if (null != dll.match(/^php7/)) {
+        } else if (null != dll.match(/^php[78]/)) {
             configure_module_dirname = configure_module_dirname + "\\src\\php7";
         } else {
             ERROR("Cannot match any known PHP version with '" + dll + "'");
diff --git a/pecl-compat/compat.h b/pecl-compat/compat.h
index c9c9ac5..d065a03 100644
--- a/pecl-compat/compat.h
+++ b/pecl-compat/compat.h
@@ -76,11 +76,7 @@
 #	include <sys/stat.h>
 #endif
 
-#ifdef PHP_WIN32
-#include <win32/php_stdint.h>
-#else
 #include <inttypes.h>
-#endif
 
 #if ZEND_EXTENSION_API_NO >= PHP_5_6_X_API_NO
 #include "zend_virtual_cwd.h"

```

**Comment by remicollet** [comment]
@Jan-E thanks, your patch is applied

**Comment by 0mars** [comment]
Thanks a lot <3

**Comment by Jan-E** [comment]
@remicollet @0mars I will have to check if the inttypes change in pecl-compat/compat.h does not break compilation for older VC-versions. It probably does. Do not merge yet.

**Comment by remicollet** [comment]
> @remicollet @0mars I will have to check if the inttypes change in pecl-compat/compat.h does not break compilation for older VC-versions. It probably does. Do not merge yet.

Others commits have been applied, please open another PR when changes for Windows will be ready

**Pull request closed without merging**